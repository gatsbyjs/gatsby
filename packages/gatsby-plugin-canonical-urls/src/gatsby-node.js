exports.pluginOptionsSchema = ({ Joi }) =>
  Joi.object({
    siteUrl: Joi.string()
      .required()
      .description(`The full URL for the site e.g. https://www.example.com`),
    stripQueryString: Joi.boolean().description(
      `Enables stripQueryString to strip query strings from paths e.g. /blog?tag=foobar becomes /blog.`
    ),
    excludes: Joi.array()
      .items(Joi.any())
      .default([])
      .description(
        `An array of paths to exclude from receiving a canonical link e.g. /blog/foobar`
      ),
  })
