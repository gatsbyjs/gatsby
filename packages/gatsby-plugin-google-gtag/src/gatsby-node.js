// @ts-check

/**
 * @type {import('gatsby').GatsbyNode["pluginOptionsSchema"]}
 */
exports.pluginOptionsSchema = ({ Joi }) =>
  Joi.object({
    trackingIds: Joi.array()
      .items(Joi.string())
      .description(
        `The tracking IDs; the tracking code won't be generated without them.`
      )
      .required(),
    gtagConfig: Joi.object()
      .keys({
        optimize_id: Joi.string().description(
          `Enable if you need to use Google Optimize.`
        ),
        anonymize_ip: Joi.boolean()
          .description(`Enable if you need to use the "anonymizeIP" function.`)
          .default(false),
      })
      .unknown(true)
      .description(
        `This object gets passed directly to the gtag config command.`
      )
      .default({}),
    pluginConfig: Joi.object()
      .keys({
        head: Joi.boolean()
          .description(
            `Puts tracking script in the <head> instead of the <body>`
          )
          .default(false),
        respectDNT: Joi.boolean()
          .description(
            `If you enable this optional option, Google Global Site Tag will not be loaded at all for visitors that have "Do Not Track" enabled.`
          )
          .default(false),
        exclude: Joi.array()
          .items(Joi.string())
          .description(
            `If you need to exclude any path from the tracking system, you can add it (one or more) to this optional array as glob expressions.`
          )
          .default([]),
        origin: Joi.string()
          .description(`Your optional self hosted origin for the script.`)
          .default(`https://www.googletagmanager.com`),
        delayOnRouteUpdate: Joi.number()
          .description(`Delay processing pageview events on route update`)
          .default(0),
        disabled: Joi.boolean()
          .description(`Prevents the tag to appear`)
          .default(false),
      })
      .description(`Configure the plugin's behavior.`),
  })
