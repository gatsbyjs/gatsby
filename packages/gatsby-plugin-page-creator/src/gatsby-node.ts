import _ from "lodash"
import glob from "globby"
import systemPath from "path"
import { sync as existsSync } from "fs-exists-cached"
import {
  CreatePagesArgs,
  ParentSpanPluginArgs,
  SetFieldsOnGraphQLNodeTypeArgs,
  PluginOptions,
  PluginCallback,
} from "gatsby"
import { parse, GraphQLString } from "gatsby/graphql"
import {
  createPath,
  watchDirectory,
  applyTrailingSlashOption,
} from "gatsby-page-utils"
import { Options as ISlugifyOptions } from "@sindresorhus/slugify"
import { createPage } from "./create-page-wrapper"
import { collectionExtractQueryString } from "./collection-extract-query-string"
import { derivePath } from "./derive-path"
import { validatePathQuery } from "./validate-path-query"
import { CODES, ERROR_MAP, prefixId } from "./error-utils"
import { createPagesFromChangedNodes } from "./create-pages-from-changed-nodes"
import type { IOptions } from "./types"
import {
  getPluginInstance,
  ICreateAPageFromNodeArgs,
} from "./tracked-nodes-state"
import { findCollectionPageFiles } from "./path-utils"
import { getCollectionRouteParams } from "./get-collection-route-params"
import { reverseLookupParams } from "./extract-query"
import { getMatchPath } from "gatsby-core-utils/match-path"

const knownCollections = new Map()

export function createPages(_: CreatePagesArgs, pluginOptions: IOptions): void {
  const instance = getPluginInstance(pluginOptions)
  if (instance.syncPages) {
    instance.syncPages()
  }
}

// Path creator.
// Auto-create pages.
// algorithm is glob /pages directory for js/jsx/cjsx files *not*
// underscored. Then create url w/ our path algorithm *unless* user
// takes control of that page component in gatsby-node.
export async function createPagesStatefully(
  {
    store,
    actions,
    reporter,
    graphql,
    emitter,
  }: CreatePagesArgs & {
    traceId: "initial-createPages"
  },
  pluginOptions: IOptions,
  doneCb: PluginCallback
): Promise<void> {
  const {
    path: pagesPath,
    pathCheck = true,
    ignore,
    slugify: slugifyOptions,
  } = pluginOptions
  try {
    const { deletePage } = actions
    const { program, config } = store.getState()
    const { trailingSlash = `always` } = config

    const exts = program.extensions.map(e => `${e.slice(1)}`).join(`,`)

    if (!pagesPath) {
      reporter.panic({
        id: prefixId(CODES.RequiredPath),
        context: {
          sourceMessage: `"path" is a required option for gatsby-plugin-page-creator

See docs here - https://www.gatsbyjs.com/plugins/gatsby-plugin-page-creator/`,
        },
      })
    }

    // Validate that the path exists.
    if (pathCheck && !existsSync(pagesPath)) {
      reporter.panic({
        id: prefixId(CODES.NonExistingPath),
        context: {
          sourceMessage: `The path passed to gatsby-plugin-page-creator does not exist on your file system:

${pagesPath}

Please pick a path to an existing directory.`,
        },
      })
    }

    const pagesDirectory = systemPath.resolve(process.cwd(), pagesPath)
    const pagesGlob = `**/*.{${exts}}`

    // Get initial list of files.
    const files = await glob(pagesGlob, { cwd: pagesPath })
    files.forEach(file => {
      createPage(
        file,
        pagesDirectory,
        actions,
        graphql,
        reporter,
        trailingSlash,
        pagesPath,
        ignore,
        slugifyOptions
      )
    })

    const knownFiles = new Set(files)
    const pluginInstance = getPluginInstance({ path: pagesPath })

    pluginInstance.syncPages = function syncPages(): void {
      createPagesFromChangedNodes({ actions, pluginInstance }, pluginOptions)
    }

    pluginInstance.resolveFields = async function resolveFields(
      nodeIds: Array<string>,
      absolutePath: string
    ): Promise<Array<Record<string, Record<string, unknown>>>> {
      const queryString = collectionExtractQueryString(
        absolutePath,
        reporter,
        nodeIds
      )
      if (!queryString) {
        return []
      }

      const { data } = await graphql<{
        nodes: Record<string, unknown>
      }>(queryString)

      if (!data) {
        return []
      }

      const nodes = Object.values(Object.values(data)[0])[0] as any as Array<
        Record<string, Record<string, unknown>>
      >
      return nodes
    }

    pluginInstance.getPathFromAResolvedNode =
      async function getPathFromAResolvedNode({
        node,
        absolutePath,
      }: ICreateAPageFromNodeArgs): Promise<string> {
        const filePath = systemPath.relative(pluginOptions.path, absolutePath)

        // URL path for the component and node
        const { derivedPath } = await derivePath(
          filePath,
          node,
          reporter,
          slugifyOptions
        )

        const hasTrailingSlash = derivedPath.endsWith(`/`)
        const path = createPath(derivedPath, hasTrailingSlash, true)
        const modifiedPath = applyTrailingSlashOption(path, trailingSlash)
        return modifiedPath
      }

    pluginInstance.createAPageFromNode = async function createAPageFromNode({
      node,
      absolutePath,
    }: ICreateAPageFromNodeArgs): Promise<
      undefined | { errors: number; path: string }
    > {
      const filePath = systemPath.relative(pluginOptions.path, absolutePath)

      const contentFilePath = node.internal?.contentFilePath
      // URL path for the component and node
      const { derivedPath, errors } = await derivePath(
        filePath,
        node,
        reporter,
        slugifyOptions
      )

      const hasTrailingSlash = derivedPath.endsWith(`/`)
      const path = createPath(derivedPath, hasTrailingSlash, true)
      const modifiedPath = applyTrailingSlashOption(path, trailingSlash)

      // We've already created a page with this path
      if (this.knownPagePaths.has(modifiedPath)) {
        return undefined
      }
      this.knownPagePaths.add(modifiedPath)
      // Params is supplied to the FE component on props.params
      const params = getCollectionRouteParams(createPath(filePath), path)
      // nodeParams is fed to the graphql query for the component
      const nodeParams = reverseLookupParams(node, absolutePath)
      // matchPath is an optional value. It's used if someone does a path like `{foo}/[bar].js`
      const matchPath = getMatchPath(path)

      const componentPath = contentFilePath
        ? `${absolutePath}?__contentFilePath=${contentFilePath}`
        : absolutePath

      actions.createPage({
        path: modifiedPath,
        matchPath,
        component: componentPath,
        context: {
          ...nodeParams,
          __params: params,
        },
      })

      const nodeId = node.id as unknown as string
      if (nodeId) {
        let templatesToPagePath = this.nodeIdToPagePath.get(nodeId)
        if (!templatesToPagePath) {
          templatesToPagePath = new Map<string, string>()
          this.nodeIdToPagePath.set(nodeId, templatesToPagePath)
        }

        templatesToPagePath.set(componentPath, modifiedPath)
      }

      return { errors, path }
    }

    pluginInstance.deletePagesCreateFromNode =
      function deletePagesCreateFromNode(id: string): void {
        const templatesToPagePaths = this.nodeIdToPagePath.get(id)
        if (templatesToPagePaths) {
          for (const [
            componentPath,
            pagePath,
          ] of templatesToPagePaths.entries()) {
            actions.deletePage({
              path: pagePath,
              component: componentPath,
            })
            this.knownPagePaths.delete(pagePath)
          }
          this.nodeIdToPagePath.delete(id)
        }
      }

    watchDirectory(
      pagesPath,
      pagesGlob,
      addedPath => {
        try {
          if (!knownFiles.has(addedPath)) {
            createPage(
              addedPath,
              pagesDirectory,
              actions,
              graphql,
              reporter,
              trailingSlash,
              pagesPath,
              ignore,
              slugifyOptions
            )
            knownFiles.add(addedPath)
          }
        } catch (e) {
          reporter.panic({
            id: prefixId(CODES.FileSystemAdd),
            context: {
              sourceMessage: e.message,
            },
          })
        }
      },
      removedPath => {
        // Delete the page for the now deleted component.
        try {
          const componentPath = systemPath.join(pagesDirectory, removedPath)
          store.getState().pages.forEach(page => {
            if (page.component === componentPath) {
              deletePage({
                path: page.path,
                component: componentPath,
              })
            }
          })
          knownFiles.delete(removedPath)

          pluginInstance.templateFileRemoved(componentPath)
        } catch (e) {
          reporter.panic({
            id: prefixId(CODES.FileSystemRemove),
            context: {
              sourceMessage: e.message,
            },
          })
        }
      }
    ).then(() => doneCb(null, null))

    emitter.on(`DELETE_NODE`, action => {
      if (action.payload?.id) {
        if (pluginInstance.trackedTypes.has(action.payload?.internal?.type)) {
          pluginInstance.changedNodesSinceLastPageCreation.deleted.set(
            action.payload.id,
            {
              id: action.payload.id,
              contentDigest: action.payload.internal.contentDigest,
            }
          )
        }
      }
    })

    emitter.on(`CREATE_NODE`, action => {
      if (pluginInstance.trackedTypes.has(action.payload?.internal?.type)) {
        // If this node was deleted before being recreated, remove it from the deleted node list
        pluginInstance.changedNodesSinceLastPageCreation.deleted.delete(
          action.payload.id
        )

        pluginInstance.changedNodesSinceLastPageCreation.created.set(
          action.payload.id,
          {
            id: action.payload.id,
            contentDigest: action.payload.internal.contentDigest,
            type: action.payload.internal.type,
          }
        )
      }
    })
  } catch (e) {
    reporter.panicOnBuild({
      id: prefixId(CODES.Generic),
      context: {
        sourceMessage: e.message,
      },
    })
  }
}

export function setFieldsOnGraphQLNodeType(
  { getNode, type, store, reporter }: SetFieldsOnGraphQLNodeTypeArgs,
  { slugify: slugifyOptions }: PluginOptions & { slugify: ISlugifyOptions }
): Record<string, unknown> {
  try {
    const extensions = store.getState().program.extensions
    const { trailingSlash = `always` } = store.getState().config
    const collectionQuery = _.camelCase(`all ${type.name}`)
    if (knownCollections.has(collectionQuery)) {
      return {
        gatsbyPath: {
          type: GraphQLString,
          args: {
            filePath: {
              type: GraphQLString,
            },
          },
          resolve: async (
            source: Record<string, unknown>,
            { filePath }: { filePath: string },
            context
          ): Promise<string> => {
            // This is a quick hack for attaching parents to the node.
            // This may be an incomprehensive fixed for the general use case
            // of connecting nodes together. However, I don't quite know how to
            // fully understand the use-cases. So this is a simple fix for this
            // one common-use, and we'll iterate as we understand.
            const sourceCopy = { ...source }
            // @ts-ignore
            if (typeof source.parent === `string`) {
              // @ts-ignore
              sourceCopy.parent = getNode(source.parent)
            }

            const getFieldValue = context.nodeModel.getFieldValue

            validatePathQuery(filePath, extensions)
            const { derivedPath } = await derivePath(
              filePath,
              sourceCopy,
              reporter,
              slugifyOptions,
              getFieldValue
            )

            const hasTrailingSlash = derivedPath.endsWith(`/`)
            const path = createPath(derivedPath, hasTrailingSlash, true)
            const modifiedPath = applyTrailingSlashOption(path, trailingSlash)

            return modifiedPath
          },
        },
      }
    }

    return {}
  } catch (e) {
    reporter.panicOnBuild({
      id: prefixId(CODES.GraphQLResolver),
      context: {
        sourceMessage: e.message,
      },
    })
    return {}
  }
}

export async function onPluginInit(
  { reporter }: ParentSpanPluginArgs,
  { path: pagesPath }: IOptions
): Promise<void> {
  if (reporter.setErrorMap) {
    reporter.setErrorMap(ERROR_MAP)
  }

  try {
    const files = await findCollectionPageFiles(pagesPath)

    await Promise.all(
      files.map(async relativePath => {
        const absolutePath = require.resolve(
          systemPath.join(pagesPath, relativePath)
        )
        const queryString = await collectionExtractQueryString(
          absolutePath,
          reporter
        )
        if (!queryString) return
        const ast = parse(queryString)
        knownCollections.set(
          // @ts-ignore
          ast.definitions[0].selectionSet.selections[0].name.value,
          relativePath
        )
      })
    )
  } catch (e) {
    reporter.panicOnBuild({
      id: prefixId(CODES.Generic),
      context: {
        sourceMessage: e.message,
      },
    })
  }
}
