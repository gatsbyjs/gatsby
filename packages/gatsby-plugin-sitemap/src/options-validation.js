import { parse as parseGraphql } from "gatsby/graphql"
import { stripIndent, oneLine } from "common-tags"

import {
  resolveSiteUrl,
  resolvePagePath,
  resolvePages,
  serialize,
  defaultFilterPages,
} from "./internals"

export const pluginOptionsSchema = ({ Joi }) =>
  Joi.object({
    plugins: Joi.array().strip(),
    output: Joi.string()
      .default(`/`)
      .description(`Folder path where sitemaps are stored in \`public\`.`),
    createLinkInHead: Joi.boolean()
      .default(true)
      .description(
        `Whether to populate the \`<head>\` of your site with a link to the sitemap.`
      ),
    entryLimit: Joi.number()
      .default(45000) // default based on upstream "sitemap" plugin default, may need optimization
      .description(
        `Number of entries per sitemap file, a sitemap index and multiple sitemaps are created if you have more entries.`
      ),
    query: Joi.string()
      .default(
        oneLine`{
          site {
            siteMetadata {
              siteUrl
            }
          }

          allSitePage {
            nodes {
              path
            }
          }
        }`
      )
      .external(pluginOptions => {
        const query = pluginOptions?.query
        if (query) {
          try {
            parseGraphql(query)
          } catch (e) {
            throw new Error(
              stripIndent`
        Invalid plugin options for "gatsby-plugin-sitemap":
        "query" must be a valid GraphQL query. Received the error "${e.message}"`
            )
          }
        }
      })
      .description(
        stripIndent`
        (GraphQL Query) The query for the data you need to generate the sitemap.
        It's required to get the site's URL, if you are not fetching it from \`site.siteMetadata.siteUrl\`,
        you will need to set a custom \`resolveSiteUrl\` function.
        If you override the query, you may need to pass in a custom \`resolvePagePath\` or
        \`resolvePages\` to keep everything working.
        If you fetch pages without using \`allSitePage.nodes\` query structure
        you will definately need to customize the \`resolvePages\` function.`
      ),
    excludes: Joi.array()
      .items(Joi.any())
      .default([])
      .description(
        stripIndent`
        An array of paths to exclude from the sitemap.
        While this is usually an array of strings it is possible to
        enter other data types into this array for custom filtering.
        Doing so will require customization of the \`filterPages\` function.`
      ),
    resolveSiteUrl: Joi.function()
      .default(() => resolveSiteUrl)
      .description(
        `Takes the output of the data query and lets you return the site URL.`
      ),
    resolvePagePath: Joi.function()
      .default(() => resolvePagePath)
      .description(
        `Takes a page object and returns the uri of the page (no domain or protocol).`
      ),
    resolvePages: Joi.function()
      .default(() => resolvePages)
      .description(
        `Takes the output of the data query and expects an array of page objects to be returned.`
      ),
    filterPages: Joi.function()
      .default(() => defaultFilterPages)
      .description(
        stripIndent`Takes the current page a string (or other object)
        from the \`exclude\` array and expects a boolean to be returned.
        \`true\` excludes the path, \`false\` keeps it.`
      ),
    serialize: Joi.function()
      .default(() => serialize)
      .description(
        `Takes the output of \`filterPages\` and lets you return a sitemap entry.`
      ),
    ignoreSitemapPathPrefix: Joi.boolean()
      .default(false)
      .description(
        `Whether to ignores the path prefix when setting the sitemap public path.`
      ),
  })
