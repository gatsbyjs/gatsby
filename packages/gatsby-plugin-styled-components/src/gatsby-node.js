// Add Babel plugin
try {
  require.resolve(`babel-plugin-styled-components`)
} catch (e) {
  throw new Error(
    `'babel-plugin-styled-components' is not installed which is needed by plugin 'gatsby-plugin-styled-components'`
  )
}

exports.pluginOptionsSchema = ({ Joi }) =>
  Joi.object({
    displayName: Joi.boolean()
      .default(true)
      .description(
        `This option enhances the attached CSS class name on each component with richer output to help identify your components in the DOM without React DevTools. In your page source you'll see: <button class="Button-asdf123 asdf123" /> instead of just <button class="asdf123" />`
      ),
    fileName: Joi.boolean()
      .default(true)
      .description(`Prefix the displayName of a component with the filename.`),
    minify: Joi.boolean()
      .default(true)
      .description(`Remove the whitespace from the CSS.`),
    namespace: Joi.string()
      .default(``)
      .description(
        `The namespace will ensure that your class names will be unique; this setting is handy when you are working with micro frontends where class name collision can occur.`
      ),
    transpileTemplateLiterals: Joi.boolean()
      .default(true)
      .description(`Transpile tagged template literals into optimized code.`),
    topLevelImportPaths: Joi.array()
      .default([])
      .items(Joi.string())
      .description(`Top level import paths allowed to identify library`),
    pure: Joi.boolean()
      .default(false)
      .description(
        `By default minifiers cannot properly perform dead code elimination on styled components because they are assumed to have side effects. This enables "pure annotations" to tell the compiler that they do not have side effects.`
      ),
    enableVendorPrefixes: Joi.boolean()
      .default(true)
      .description(`Enables vendor prefixing`),
  })

exports.onCreateBabelConfig = ({ stage, actions }, pluginOptions) => {
  const ssr = stage === `build-html` || stage === `build-javascript`
  const { enableVendorPrefixes: _, ...babelOptions } = pluginOptions

  actions.setBabelPlugin({
    name: `babel-plugin-styled-components`,
    stage,
    options: { ...babelOptions, ssr },
  })
}
