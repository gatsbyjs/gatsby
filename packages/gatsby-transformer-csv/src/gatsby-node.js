const csv = require(`csvtojson`)
const _ = require(`lodash`)
const { typeNameFromFile } = require(`./index`)

const convertToJson = (data, options) =>
  csv(options)
    .fromString(data)
    .then(jsonData => jsonData, new Error(`CSV to JSON conversion failed!`))

function shouldOnCreateNode({ node }, pluginOptions = {}) {
  const { extension } = node
  const { extensions } = pluginOptions

  return extensions ? extensions.includes(extension) : extension === `csv`
}

const typeCache = new Map()

async function onCreateNode(
  { node, actions, loadNodeContent, createNodeId, createContentDigest },
  pluginOptions
) {
  const { createNode, createParentChildLink } = actions

  // Destructure out our custom options
  const { typeName, nodePerFile, ...options } = pluginOptions || {}

  // Load file contents
  const content = await loadNodeContent(node)

  // Parse
  const parsedContent = await convertToJson(content, options)

  // Generate the type
  function getType({ node, object }) {
    if (pluginOptions && _.isFunction(typeName)) {
      return pluginOptions.typeName({ node, object })
    } else if (pluginOptions && _.isString(typeName)) {
      if (typeCache.has(node.internal.type)) {
        return typeCache.get(node.internal.type)
      } else {
        const type = _.upperFirst(_.camelCase(typeName))
        typeCache.set(node.internal.type, type)
        return type
      }
    } else {
      return typeNameFromFile({ node })
    }
  }

  // Generate the new node
  function transformObject(obj, i) {
    const csvNode = {
      ...obj,
      id:
        obj.id ??
        createNodeId(`${node.id} [${i}] >>> ${node.extension.toUpperCase()}`),
      children: [],
      parent: node.id,
      internal: {
        contentDigest: createContentDigest(obj),
        // TODO make choosing the "type" a lot smarter. This assumes
        // the parent node is a file.
        // PascalCase
        type: getType({ node, object: parsedContent }),
      },
    }

    createNode(csvNode)
    createParentChildLink({ parent: node, child: csvNode })
  }

  async function transformArrayChunk({ chunk, startCount }) {
    for (let i = 0, l = chunk.length; i < l; i++) {
      const obj = chunk[i]
      transformObject(obj, i + startCount)
      await new Promise(resolve =>
        setImmediate(() => {
          resolve()
        })
      )
    }
  }

  if (_.isArray(parsedContent)) {
    if (pluginOptions && nodePerFile) {
      if (pluginOptions && _.isString(nodePerFile)) {
        transformObject({ [nodePerFile]: parsedContent }, 0)
      } else {
        transformObject({ items: parsedContent }, 0)
      }
    } else {
      const chunks = _.chunk(parsedContent, 100)
      let count = 0
      for (const chunk of chunks) {
        await transformArrayChunk({ chunk, startCount: count })
        count += chunk.length
      }
    }
  }

  return
}

exports.shouldOnCreateNode = shouldOnCreateNode
exports.onCreateNode = onCreateNode
