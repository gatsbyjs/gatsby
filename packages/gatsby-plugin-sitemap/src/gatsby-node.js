import path from "path"
import { simpleSitemapAndIndex } from "sitemap"
import { pageFilter, prefixPath, REPORTER_PREFIX } from "./internals"
import { pluginOptionsSchema } from "./options-validation"

exports.pluginOptionsSchema = pluginOptionsSchema

exports.onPostBuild = async (
  { graphql, reporter, basePath, pathPrefix },
  {
    output,
    entryLimit,
    query,
    excludes,
    resolveSiteUrl,
    resolvePagePath,
    resolvePages,
    filterPages,
    serialize,
    ignoreSitemapPathPrefix,
  }
) => {
  const { data: queryRecords, errors } = await graphql(query)

  // resolvePages and resolveSiteUrl are allowed to be sync or async. The Promise.resolve handles each possibility
  const siteUrl = await Promise.resolve(resolveSiteUrl(queryRecords)).catch(
    err => reporter.panic(`${REPORTER_PREFIX} Error resolving Site URL`, err)
  )

  if (errors) {
    reporter.panic(
      `Error executing the GraphQL query inside gatsby-plugin-sitemap:\n`,
      errors
    )
  }

  const allPages = await Promise.resolve(resolvePages(queryRecords)).catch(
    err => reporter.panic(`${REPORTER_PREFIX} Error resolving Pages`, err)
  )

  if (!Array.isArray(allPages)) {
    reporter.panic(
      `${REPORTER_PREFIX} The \`resolvePages\` function did not return an array.`
    )
  }

  reporter.verbose(
    `${REPORTER_PREFIX} Filtering ${allPages.length} pages based on ${excludes.length} excludes`
  )

  const { filteredPages, messages } = pageFilter(
    {
      allPages,
      filterPages,
      excludes,
    },
    { reporter }
  )

  messages.forEach(message => reporter.verbose(message))

  reporter.verbose(
    `${REPORTER_PREFIX} ${filteredPages.length} pages remain after filtering`
  )

  const serializedPages = []

  for (const page of filteredPages) {
    try {
      const { url, ...rest } = await Promise.resolve(
        serialize(page, { resolvePagePath })
      )
      serializedPages.push({
        url: prefixPath({ url, siteUrl, pathPrefix: basePath }),
        ...rest,
      })
    } catch (err) {
      reporter.panic(`${REPORTER_PREFIX} Error serializing pages`, err)
    }
  }

  const sitemapWritePath = path.join(`public`, output)
  const sitemapPublicPath = path.posix.join(
    ignoreSitemapPathPrefix ? basePath : pathPrefix,
    output
  )

  return simpleSitemapAndIndex({
    hostname: siteUrl,
    publicBasePath: sitemapPublicPath,
    destinationDir: sitemapWritePath,
    sourceData: serializedPages,
    limit: entryLimit,
    gzip: false,
  })
}
