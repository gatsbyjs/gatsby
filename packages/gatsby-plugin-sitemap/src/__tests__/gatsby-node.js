import { Joi } from "gatsby-plugin-utils"
import path from "path"
import sitemap from "sitemap"
import { onPostBuild } from "../gatsby-node"
import { pluginOptionsSchema } from "../options-validation"

jest.mock(`sitemap`, () => {
  return {
    simpleSitemapAndIndex: jest.fn(),
  }
})

const schema = pluginOptionsSchema({ Joi })

const pathPrefix = ``

const reporter = {
  verbose: jest.fn(),
  panic: jest.fn(),
}

beforeEach(() => {
  global.__PATH_PREFIX__ = ``

  sitemap.simpleSitemapAndIndex.mockReset()
})

describe(`gatsby-plugin-sitemap Node API`, () => {
  it(`should succeed with default options`, async () => {
    const graphql = jest.fn()
    graphql.mockResolvedValue({
      data: {
        site: {
          siteMetadata: {
            siteUrl: `http://dummy.url`,
          },
        },
        allSitePage: {
          nodes: [
            {
              path: `/page-1`,
            },
            {
              path: `/page-2`,
            },
          ],
        },
      },
    })
    await onPostBuild(
      { graphql, pathPrefix, reporter },
      await schema.validateAsync({})
    )
    const { destinationDir, sourceData } =
      sitemap.simpleSitemapAndIndex.mock.calls[0][0]
    expect(destinationDir).toEqual(path.join(`public`, `/`))
    expect(sourceData).toMatchSnapshot()
  })

  it(`should accept a custom query`, async () => {
    const graphql = jest.fn()
    const siteUrl = `http://dummy.url`
    graphql.mockResolvedValue({
      data: {
        allSitePage: {
          edges: [
            {
              node: {
                path: `/page-1`,
              },
            },
            {
              node: {
                path: `/post/exclude-page`,
              },
            },
          ],
        },
      },
    })
    const customQuery = `
      {
        allSitePage {
          edges {
            node {
              path
            }
          }
        }
    }`
    const options = {
      output: `custom-folder`,
      resolveSiteUrl: () => siteUrl,
      resolvePages: data => data.allSitePage.edges.map(edge => edge.node),
      serialize: (page, { resolvePagePath }) => {
        return {
          url: resolvePagePath(page),
          changefreq: `weekly`,
          priority: 0.8,
        }
      },
      excludes: [`/post/exclude-page`],
      query: customQuery,
    }

    await onPostBuild(
      { graphql, pathPrefix, reporter },
      await schema.validateAsync(options)
    )

    const { destinationDir, sourceData } =
      sitemap.simpleSitemapAndIndex.mock.calls[0][0]

    expect(destinationDir).toEqual(path.join(`public`, `custom-folder`))
    expect(sourceData).toMatchSnapshot()
    expect(graphql).toBeCalledWith(customQuery)
  })

  it(`should include path prefix when creating index sitemap`, async () => {
    const graphql = jest.fn()
    graphql.mockResolvedValue({
      data: {
        site: {
          siteMetadata: {
            siteUrl: `http://dummy.url`,
          },
        },
        allSitePage: {
          nodes: [
            {
              path: `/page-1`,
            },
            {
              path: `/page-2`,
            },
          ],
        },
      },
    })

    const options = {
      entryLimit: 1,
    }
    const assetPrefix = `https://cdn.example.com`
    const prefix = `/test`
    await onPostBuild(
      {
        graphql,
        basePath: prefix,
        pathPrefix: `${assetPrefix}${prefix}`,
        reporter,
      },
      await schema.validateAsync(options)
    )
    const { sourceData } = sitemap.simpleSitemapAndIndex.mock.calls[0][0]
    expect(sourceData).toMatchSnapshot()
  })

  it(`should ignore path prefix for sitemap public path when creating index sitemap with ignoreSitemapPathPrefix`, async () => {
    const graphql = jest.fn()
    graphql.mockResolvedValue({
      data: {
        site: {
          siteMetadata: {
            siteUrl: `http://dummy.url`,
          },
        },
        allSitePage: {
          nodes: [
            {
              path: `/page-1`,
            },
            {
              path: `/page-2`,
            },
          ],
        },
      },
    })
    const prefix = `/test`
    const subdir = `/subdir`
    const options = {
      output: subdir,
      ignoreSitemapPathPrefix: true,
    }
    const assetPrefix = `https://cdn.example.com`
    await onPostBuild(
      {
        graphql,
        basePath: prefix,
        pathPrefix: `${assetPrefix}${prefix}`,
        reporter,
      },
      await schema.validateAsync(options)
    )
    const { publicBasePath } = sitemap.simpleSitemapAndIndex.mock.calls[0][0]
    expect(publicBasePath).toEqual(`${prefix}${subdir}`)
  })

  it(`should output modified paths to sitemap`, async () => {
    const graphql = jest.fn()
    graphql.mockResolvedValue({
      data: {
        site: {
          siteMetadata: {
            siteUrl: `http://dummy.url`,
          },
        },
        allSitePage: {
          nodes: [
            {
              path: `/page-1`,
            },
            {
              path: `/page-2`,
            },
          ],
        },
      },
    })
    const prefix = `/test`
    const subdir = `/subdir`
    const options = {
      output: subdir,
    }
    await onPostBuild(
      {
        graphql,
        basePath: prefix,
        pathPrefix: prefix,
        reporter,
      },
      await schema.validateAsync(options)
    )
    expect(sitemap.simpleSitemapAndIndex.mock.calls[0][0].publicBasePath).toBe(
      path.posix.join(prefix, subdir)
    )
    expect(sitemap.simpleSitemapAndIndex.mock.calls[0][0].destinationDir).toBe(
      path.join(`public`, subdir)
    )
  })

  it(`should output modified paths to sitemap with asset prefix`, async () => {
    const graphql = jest.fn()
    graphql.mockResolvedValue({
      data: {
        site: {
          siteMetadata: {
            siteUrl: `http://dummy.url`,
          },
        },
        allSitePage: {
          nodes: [
            {
              path: `/page-1`,
            },
            {
              path: `/page-2`,
            },
          ],
        },
      },
    })
    const prefix = `/test`
    const subdir = `/subdir`
    const options = {
      output: subdir,
    }
    const assetPrefix = `https://cdn.example.com`
    await onPostBuild(
      {
        graphql,
        basePath: prefix,
        pathPrefix: `${assetPrefix}${prefix}`,
        reporter,
      },
      await schema.validateAsync(options)
    )
    expect(sitemap.simpleSitemapAndIndex.mock.calls[0][0].publicBasePath).toBe(
      path.posix.join(assetPrefix, prefix, subdir)
    )
    expect(sitemap.simpleSitemapAndIndex.mock.calls[0][0].destinationDir).toBe(
      path.join(`public`, subdir)
    )
  })
})
