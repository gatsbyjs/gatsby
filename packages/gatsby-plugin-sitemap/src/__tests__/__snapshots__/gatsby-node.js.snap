// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-sitemap Node API should accept a custom query 1`] = `
Array [
  Object {
    "changefreq": "weekly",
    "priority": 0.8,
    "url": "http://dummy.url/page-1",
  },
]
`;

exports[`gatsby-plugin-sitemap Node API should include path prefix when creating index sitemap 1`] = `
Array [
  Object {
    "changefreq": "daily",
    "priority": 0.7,
    "url": "http://dummy.url/test/page-1",
  },
  Object {
    "changefreq": "daily",
    "priority": 0.7,
    "url": "http://dummy.url/test/page-2",
  },
]
`;

exports[`gatsby-plugin-sitemap Node API should succeed with default options 1`] = `
Array [
  Object {
    "changefreq": "daily",
    "priority": 0.7,
    "url": "http://dummy.url/page-1",
  },
  Object {
    "changefreq": "daily",
    "priority": 0.7,
    "url": "http://dummy.url/page-2",
  },
]
`;
